File: lib/features/imports/csv/csv_row_converter.dart
/// One CSV row → the Health Connect records it represents.
///
/// Pure and synchronous: no repository, no clock, no I/O. Everything the
/// conversion needs is the row, the mapping, and the catalog.
library;

import '../../../domain/model/apple_health_import_records.dart';
import '../../../domain/model/import_client_record_id.dart';
import 'csv_column_mapping.dart';
import 'csv_datetime_format.dart';
import 'csv_import_metric.dart';
import 'csv_import_models.dart';
import 'csv_table_reader.dart';

/// The namespace every CSV-imported record's `clientRecordId` carries.
///
/// Distinct from the Apple importer's so the two can never collide on the same
/// id and silently overwrite each other's records.
const String kCsvClientRecordIdNamespace = 'csv';

/// What one row produced.
class CsvRowConversion {
  const CsvRowConversion({
    this.records = const [],
    this.diagnostics = const [],
  });

  final List<ImportRecord> records;
  final List<CsvImportDiagnostic> diagnostics;
}

/// Converts [row] under [mapping].
///
/// Failure granularity matches the Apple importer: a bad **timestamp** or a row
/// too short costs the whole row, because nothing in it can be placed in time; a
/// bad **value** costs only that metric, so one unparsable body-fat cell does
/// not throw away a perfectly good weight.
CsvRowConversion convertCsvRow({
  required CsvRow row,
  required CsvImportMapping mapping,
}) {
  final timestampColumn = mapping.timestampColumn;
  if (timestampColumn == null) return const CsvRowConversion();

  final metricColumns = mapping.metricColumns;
  if (metricColumns.isEmpty) return const CsvRowConversion();

  // A row is "too short" only relative to what the mapping actually reads.
  // Trailing empty columns are normal in exports and must not reject the row.
  final highestIndex = [
    timestampColumn.columnIndex,
    for (final column in metricColumns) column.columnIndex,
  ].reduce((a, b) => a > b ? a : b);
  if (row.fields.length <= highestIndex) {
    return CsvRowConversion(
      diagnostics: [
        CsvImportDiagnostic(
          rowNumber: row.rowNumber,
          reason: CsvImportDiagnosticReason.wrongFieldCount,
          detail: '${row.fields.length} fields, needs ${highestIndex + 1}',
        ),
      ],
    );
  }

  final timestampText = row.cell(timestampColumn.columnIndex);
  if (timestampText == null) {
    return CsvRowConversion(
      diagnostics: [
        CsvImportDiagnostic(
          rowNumber: row.rowNumber,
          reason: CsvImportDiagnosticReason.missingTimestamp,
          columnIndex: timestampColumn.columnIndex,
        ),
      ],
    );
  }

  final instant = resolveCsvInstant(timestampText, mapping.dateTime);
  if (instant == null) {
    return CsvRowConversion(
      diagnostics: [
        CsvImportDiagnostic(
          rowNumber: row.rowNumber,
          reason: CsvImportDiagnosticReason.unparsableTimestamp,
          columnIndex: timestampColumn.columnIndex,
          detail: timestampText,
        ),
      ],
    );
  }

  // Resolved once per row, before the loop, because any number of mass-share
  // metrics can need it.
  final rowWeightKg = _resolveRowWeightKg(row, mapping);

  final records = <ImportRecord>[];
  final diagnostics = <CsvImportDiagnostic>[];

  for (final column in metricColumns) {
    final metric = column.metric!;
    final spec = kCsvMetricCatalog[metric];
    final interpretation = column.effectiveInterpretation;
    if (spec == null || interpretation == null) continue;

    final text = row.cell(column.columnIndex);
    // A blank cell is a gap in the data, not an error: scales skip metrics.
    if (text == null) continue;

    final raw = parseCsvNumber(text);
    if (raw == null) {
      diagnostics.add(
        CsvImportDiagnostic(
          rowNumber: row.rowNumber,
          reason: CsvImportDiagnosticReason.unparsableNumber,
          columnIndex: column.columnIndex,
          detail: text,
        ),
      );
      continue;
    }

    final double canonical;
    switch (interpretation) {
      case CsvDirectValue(:final unit):
        canonical = convertCsvValueToCanonical(raw, unit);
      case CsvMassShareOfWeight(:final unit):
        if (rowWeightKg == null || rowWeightKg <= 0) {
          diagnostics.add(
            CsvImportDiagnostic(
              rowNumber: row.rowNumber,
              reason: CsvImportDiagnosticReason.derivationMissingWeight,
              columnIndex: column.columnIndex,
            ),
          );
          continue;
        }
        final massKg = convertCsvValueToCanonical(raw, unit);
        canonical = massKg / rowWeightKg * 100;
    }

    if (canonical < spec.plausibleMin || canonical > spec.plausibleMax) {
      diagnostics.add(
        CsvImportDiagnostic(
          rowNumber: row.rowNumber,
          reason: CsvImportDiagnosticReason.outOfRange,
          columnIndex: column.columnIndex,
          detail: canonical.toStringAsFixed(2),
        ),
      );
      continue;
    }

    records.add(
      buildCsvImportRecord(
        metric: metric,
        value: canonical,
        instant: instant,
      ),
    );
  }

  return CsvRowConversion(records: records, diagnostics: diagnostics);
}

/// The row's body weight in kg, or null when the mapping has no weight column,
/// the cell is blank, or it does not parse.
double? _resolveRowWeightKg(CsvRow row, CsvImportMapping mapping) {
  final column = mapping.weightColumn;
  if (column == null) return null;
  final text = row.cell(column.columnIndex);
  if (text == null) return null;
  final raw = parseCsvNumber(text);
  if (raw == null) return null;
  final interpretation = column.effectiveInterpretation;
  if (interpretation is! CsvDirectValue) return null;
  return convertCsvValueToCanonical(raw, interpretation.unit);
}

/// Builds the record for [metric] at [instant] from an already-canonical [value].
ImportRecord buildCsvImportRecord({
  required CsvImportMetric metric,
  required double value,
  required CsvInstant instant,
}) {
  final targetType = kCsvMetricCatalog[metric]!.targetType;
  final clientRecordId = buildCsvClientRecordId(
    targetType: targetType,
    utc: instant.utc,
  );

  return switch (metric) {
    CsvImportMetric.weight => WeightImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        kilograms: value,
      ),
    CsvImportMetric.bodyFat => BodyFatImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        percent: value,
      ),
    CsvImportMetric.leanBodyMass => LeanBodyMassImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        kilograms: value,
      ),
    CsvImportMetric.boneMass => BoneMassImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        kilograms: value,
      ),
    CsvImportMetric.bodyWaterMass => BodyWaterMassImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        kilograms: value,
      ),
    CsvImportMetric.height => HeightImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        meters: value,
      ),
    CsvImportMetric.basalMetabolicRate => BasalMetabolicRateImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        kilocaloriesPerDay: value,
      ),
    // Health Connect models heart rate as a SERIES, not an instant. A CSV row is
    // one spot reading, so it becomes a one-sample series whose window is the
    // single instant — the same shape the Apple importer builds for a lone
    // HKQuantityTypeIdentifierHeartRate sample.
    CsvImportMetric.heartRate => HeartRateImportRecord(
        clientRecordId: clientRecordId,
        startTime: instant.utc,
        startZoneOffset: instant.offset,
        endTime: instant.utc,
        endZoneOffset: instant.offset,
        samples: [HeartRateSampleValue(instant.utc, value.round())],
      ),
    // Health Connect stores this one as an integer.
    CsvImportMetric.restingHeartRate => RestingHeartRateImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        beatsPerMinute: value.round(),
      ),
    CsvImportMetric.heartRateVariability =>
      HeartRateVariabilityRmssdImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        rmssdMillis: value,
      ),
    CsvImportMetric.oxygenSaturation => OxygenSaturationImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        percent: value,
      ),
    CsvImportMetric.respiratoryRate => RespiratoryRateImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        rate: value,
      ),
    CsvImportMetric.bodyTemperature => BodyTemperatureImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        celsius: value,
      ),
    CsvImportMetric.basalBodyTemperature => BasalBodyTemperatureImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        celsius: value,
      ),
    CsvImportMetric.bloodGlucose => BloodGlucoseImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        millimolesPerLiter: value,
      ),
    CsvImportMetric.vo2Max => Vo2MaxImportRecord(
        clientRecordId: clientRecordId,
        time: instant.utc,
        zoneOffset: instant.offset,
        vo2MillilitersPerMinuteKilogram: value,
      ),
  };
}

/// The identity of a CSV-imported record: **record type and instant, not value.**
///
/// Health Connect upserts on `clientRecordId`, so leaving the value out is what
/// makes re-importing a corrected file REPLACE the old record instead of leaving
/// two weights at the same instant. The cost is that two genuinely different
/// measurements at the identical instant collapse to one — for a scale, the right
/// trade.
///
/// Also deliberately excluded: the file name, the column header, the unit chosen
/// and the mapping. Re-exporting the same history with the columns reordered, or
/// in pounds instead of kilograms, resolves to the same records.
String buildCsvClientRecordId({
  required String targetType,
  required DateTime utc,
}) =>
    buildImportClientRecordId(
      kCsvClientRecordIdNamespace,
      targetType,
      [targetType, utc.toUtc().millisecondsSinceEpoch].join('|'),
    );

/// The canonical values [metric] would take across [rows].
///
/// Runs the REAL conversion, derivations and plausibility rejections included,
/// so the range the confirm step shows is the range that will actually be
/// written — which is the whole point of showing it. A fat-mass column divided
/// by the wrong weight column surfaces here as 3% or 150%.
List<double> previewCanonicalValues({
  required List<List<String>> rows,
  required CsvImportMapping mapping,
  required CsvImportMetric metric,
}) {
  final targetType = kCsvMetricCatalog[metric]?.targetType;
  if (targetType == null) return const [];

  final values = <double>[];
  for (var index = 0; index < rows.length; index++) {
    final conversion = convertCsvRow(
      // +2: 1-based, and past the header row.
      row: CsvRow(rowNumber: index + 2, fields: rows[index]),
      mapping: mapping,
    );
    for (final record in conversion.records) {
      if (record.targetType != targetType) continue;
      final value = canonicalValueOf(record);
      if (value != null) values.add(value);
    }
  }
  return values;
}

/// The earliest and latest wall-clock times [rows] resolve to under [mapping],
/// or null when none of them parse.
///
/// Wall clock, not the UTC instant: this is shown to the user to check against
/// the dates they can see in their own file, so it has to read the way the file
/// reads. The offset is added back for exactly that reason.
///
/// Resolved through [resolveCsvInstant], the same path the import takes, so the
/// span cannot disagree with what gets written. It is the guard the single-row
/// echo cannot be: a day/month mix-up that happens to leave row 1 plausible
/// shows up here as a span running to the wrong month, or backwards.
({DateTime first, DateTime last})? previewInstantRange({
  required List<List<String>> rows,
  required CsvImportMapping mapping,
}) {
  final column = mapping.timestampColumn;
  if (column == null) return null;

  DateTime? first;
  DateTime? last;
  for (final fields in rows) {
    if (column.columnIndex >= fields.length) continue;
    final text = fields[column.columnIndex].trim();
    if (text.isEmpty) continue;
    final instant = resolveCsvInstant(text, mapping.dateTime);
    if (instant == null) continue;
    final wallClock = instant.utc.add(instant.offset);
    if (first == null || wallClock.isBefore(first)) first = wallClock;
    if (last == null || wallClock.isAfter(last)) last = wallClock;
  }

  if (first == null || last == null) return null;
  return (first: first, last: last);
}

/// The single number a body record carries, in its canonical unit.
double? canonicalValueOf(ImportRecord record) => switch (record) {
      WeightImportRecord(:final kilograms) => kilograms,
      BodyFatImportRecord(:final percent) => percent,
      LeanBodyMassImportRecord(:final kilograms) => kilograms,
      BoneMassImportRecord(:final kilograms) => kilograms,
      BodyWaterMassImportRecord(:final kilograms) => kilograms,
      HeightImportRecord(:final meters) => meters,
      BasalMetabolicRateImportRecord(:final kilocaloriesPerDay) =>
        kilocaloriesPerDay,
      // A CSV-built heart rate always has exactly one sample.
      HeartRateImportRecord(:final samples) when samples.isNotEmpty =>
        samples.first.beatsPerMinute.toDouble(),
      RestingHeartRateImportRecord(:final beatsPerMinute) =>
        beatsPerMinute.toDouble(),
      HeartRateVariabilityRmssdImportRecord(:final rmssdMillis) => rmssdMillis,
      OxygenSaturationImportRecord(:final percent) => percent,
      RespiratoryRateImportRecord(:final rate) => rate,
      BodyTemperatureImportRecord(:final celsius) => celsius,
      BasalBodyTemperatureImportRecord(:final celsius) => celsius,
      BloodGlucoseImportRecord(:final millimolesPerLiter) => millimolesPerLiter,
      Vo2MaxImportRecord(:final vo2MillilitersPerMinuteKilogram) =>
        vo2MillilitersPerMinuteKilogram,
      _ => null,
    };

/// Parses a numeric cell, tolerating a comma decimal separator and thousands
/// separators.
///
/// A semicolon-delimited European export writes `78,4`; reading that as 784 or
/// as null would both be wrong. Only unambiguous shapes are accepted — a value
/// containing both separators is read with the LAST one as the decimal point.
double? parseCsvNumber(String text) {
  var value = text.trim();
  if (value.isEmpty) return null;

  // Strip anything that is not part of a number (stray units, currency, spaces).
  value = value.replaceAll(RegExp(r'[^0-9,.\-+eE]'), '');
  if (value.isEmpty) return null;

  final lastComma = value.lastIndexOf(',');
  final lastDot = value.lastIndexOf('.');
  if (lastComma >= 0 && lastDot >= 0) {
    if (lastComma > lastDot) {
      // 1.234,5 — dot groups, comma decides.
      value = value.replaceAll('.', '').replaceFirst(',', '.');
    } else {
      // 1,234.5 — comma groups.
      value = value.replaceAll(',', '');
    }
  } else if (lastComma >= 0) {
    // Only commas. Treat a single one as the decimal point; several means
    // thousands grouping.
    value = value.indexOf(',') == lastComma
        ? value.replaceFirst(',', '.')
        : value.replaceAll(',', '');
  }

  return double.tryParse(value);
}

File: app/src/main/kotlin/tech/mmarca/openvitals/features/imports/csv/CsvRowConverter.kt
package tech.mmarca.openvitals.features.imports.csv

import androidx.health.connect.client.records.BasalBodyTemperatureRecord
import androidx.health.connect.client.records.BasalMetabolicRateRecord
import androidx.health.connect.client.records.BloodGlucoseRecord
import androidx.health.connect.client.records.BodyFatRecord
import androidx.health.connect.client.records.BodyTemperatureRecord
import androidx.health.connect.client.records.BodyWaterMassRecord
import androidx.health.connect.client.records.BoneMassRecord
import androidx.health.connect.client.records.HeartRateRecord
import androidx.health.connect.client.records.HeartRateVariabilityRmssdRecord
import androidx.health.connect.client.records.HeightRecord
import androidx.health.connect.client.records.LeanBodyMassRecord
import androidx.health.connect.client.records.MealType
import androidx.health.connect.client.records.OxygenSaturationRecord
import androidx.health.connect.client.records.Record
import androidx.health.connect.client.records.RespiratoryRateRecord
import androidx.health.connect.client.records.RestingHeartRateRecord
import androidx.health.connect.client.records.StepsRecord
import androidx.health.connect.client.records.Vo2MaxRecord
import androidx.health.connect.client.records.WeightRecord
import androidx.health.connect.client.records.metadata.Device
import androidx.health.connect.client.records.metadata.Metadata
import androidx.health.connect.client.units.BloodGlucose
import androidx.health.connect.client.units.Length
import androidx.health.connect.client.units.Mass
import androidx.health.connect.client.units.Power
import androidx.health.connect.client.units.Temperature
import java.security.MessageDigest
import java.time.Duration
import java.time.Instant
import java.time.LocalDateTime
import java.util.Locale
import kotlin.math.roundToLong

/** One CSV row to the Health Connect records it represents. Pure: no I/O, no clock. */

/** The `clientRecordId` namespace for CSV imports. Distinct from the Apple importer's. */
const val CSV_CLIENT_RECORD_ID_NAMESPACE = "csv"

/** A record built from one CSV cell, with everything the import loop needs. */
data class CsvConvertedRecord(
    val metric: CsvImportMetric,
    /** Mirrors [CsvMetricSpec.targetType]; a segment of [clientRecordId]. */
    val targetType: String,
    val recordType: kotlin.reflect.KClass<out Record>,
    val clientRecordId: String,
    /** The measurement's instant, used to bound the duplicate-lookup window. */
    val instant: Instant,
    /** The value in the metric's canonical unit, for the confirm-step preview. */
    val canonicalValue: Double,
    val record: Record,
)

/** What one row produced. */
data class CsvRowConversion(
    val records: List<CsvConvertedRecord> = emptyList(),
    val diagnostics: List<CsvImportDiagnostic> = emptyList(),
)

/**
 * Converts [row] under [mapping]. A bad timestamp costs the row, a bad value
 * only that metric, a bad end timestamp only the interval metrics.
 */
fun convertCsvRow(
    row: CsvRow,
    mapping: CsvImportMapping,
): CsvRowConversion {
    val timestampColumn = mapping.timestampColumn ?: return CsvRowConversion()

    val metricColumns = mapping.metricColumns
    if (metricColumns.isEmpty()) return CsvRowConversion()

    // "Too short" is relative to what the mapping reads: trailing empty columns
    // are normal. The end column is not counted; a missing end falls back to
    // the one-minute span.
    val endTimestampColumn = mapping.endTimestampColumn
    val highestIndex = (metricColumns.map { it.columnIndex } + timestampColumn.columnIndex).max()
    if (row.fields.size <= highestIndex) {
        return CsvRowConversion(
            diagnostics = listOf(
                CsvImportDiagnostic(
                    rowNumber = row.rowNumber,
                    reason = CsvImportDiagnosticReason.WRONG_FIELD_COUNT,
                    detail = "${row.fields.size} fields, needs ${highestIndex + 1}",
                ),
            ),
        )
    }

    val timestampText = row.cell(timestampColumn.columnIndex)
        ?: return CsvRowConversion(
            diagnostics = listOf(
                CsvImportDiagnostic(
                    rowNumber = row.rowNumber,
                    reason = CsvImportDiagnosticReason.MISSING_TIMESTAMP,
                    columnIndex = timestampColumn.columnIndex,
                ),
            ),
        )

    val instant = resolveCsvInstant(timestampText, mapping.dateTime)
        ?: return CsvRowConversion(
            diagnostics = listOf(
                CsvImportDiagnostic(
                    rowNumber = row.rowNumber,
                    reason = CsvImportDiagnosticReason.UNPARSABLE_TIMESTAMP,
                    columnIndex = timestampColumn.columnIndex,
                    detail = timestampText,
                ),
            ),
        )

    // Resolved once per row: several mass-share metrics may need it.
    val rowWeightKg = resolveRowWeightKg(row, mapping)

    // A bad end costs only the interval metrics.
    val endText = endTimestampColumn?.let { row.cell(it.columnIndex) }
    val endInstant = endText?.let { resolveCsvInstant(it, mapping.dateTime) }

    val records = mutableListOf<CsvConvertedRecord>()
    val diagnostics = mutableListOf<CsvImportDiagnostic>()

    for (column in metricColumns) {
        val metric = column.metric!!
        val spec = CsvMetricCatalog[metric]
        val interpretation = column.effectiveInterpretation
        if (spec == null || interpretation == null) continue

        // No end named: a one-minute span. An end that is stated but garbage,
        // or not after the start, still rejects.
        var intervalEnd: CsvInstant? = null
        if (spec.isInterval) {
            val reason = when {
                endText == null -> null
                endInstant == null -> CsvImportDiagnosticReason.UNPARSABLE_END_TIMESTAMP
                !endInstant.utc.isAfter(instant.utc) -> CsvImportDiagnosticReason.END_NOT_AFTER_START
                else -> null
            }
            if (reason != null) {
                diagnostics += CsvImportDiagnostic(
                    rowNumber = row.rowNumber,
                    reason = reason,
                    columnIndex = endTimestampColumn?.columnIndex ?: column.columnIndex,
                    detail = endText,
                )
                continue
            }
            intervalEnd = endInstant ?: instant.plusDefaultIntervalSpan()
        }

        // A blank cell is a gap, not an error.
        val text = row.cell(column.columnIndex) ?: continue

        val raw = parseCsvNumber(text)
        if (raw == null) {
            diagnostics += CsvImportDiagnostic(
                rowNumber = row.rowNumber,
                reason = CsvImportDiagnosticReason.UNPARSABLE_NUMBER,
                columnIndex = column.columnIndex,
                detail = text,
            )
            continue
        }

        val canonical: Double = when (interpretation) {
            is CsvDirectValue -> convertCsvValueToCanonical(raw, interpretation.unit)
            is CsvMassShareOfWeight -> {
                if (rowWeightKg == null || rowWeightKg <= 0) {
                    diagnostics += CsvImportDiagnostic(
                        rowNumber = row.rowNumber,
                        reason = CsvImportDiagnosticReason.DERIVATION_MISSING_WEIGHT,
                        columnIndex = column.columnIndex,
                    )
                    continue
                }
                val massKg = convertCsvValueToCanonical(raw, interpretation.unit)
                massKg / rowWeightKg * 100
            }
        }

        if (canonical < spec.plausibleMin || canonical > spec.plausibleMax) {
            diagnostics += CsvImportDiagnostic(
                rowNumber = row.rowNumber,
                reason = CsvImportDiagnosticReason.OUT_OF_RANGE,
                columnIndex = column.columnIndex,
                detail = String.format(Locale.US, "%.2f", canonical),
            )
            continue
        }

        records += buildCsvImportRecord(
            metric = metric,
            value = canonical,
            instant = instant,
            end = intervalEnd,
        )
    }

    return CsvRowConversion(records = records, diagnostics = diagnostics)
}

/** The row's body weight in kg, or null. */
private fun resolveRowWeightKg(row: CsvRow, mapping: CsvImportMapping): Double? {
    val column = mapping.weightColumn ?: return null
    val text = row.cell(column.columnIndex) ?: return null
    val raw = parseCsvNumber(text) ?: return null
    val interpretation = column.effectiveInterpretation
    if (interpretation !is CsvDirectValue) return null
    return convertCsvValueToCanonical(raw, interpretation.unit)
}

/**
 * Builds the record for [metric] at [instant] from a canonical [value].
 * An interval metric also needs [end], already validated by the caller.
 */
fun buildCsvImportRecord(
    metric: CsvImportMetric,
    value: Double,
    instant: CsvInstant,
    end: CsvInstant? = null,
): CsvConvertedRecord {
    val spec = CsvMetricCatalog.getValue(metric)
    val clientRecordId = buildCsvClientRecordId(targetType = spec.targetType, utc = instant.utc)
    val metadata = csvMetadata(clientRecordId)
    val time = instant.utc
    val offset = instant.offset

    val record: Record = when (metric) {
        CsvImportMetric.WEIGHT -> WeightRecord(time, offset, Mass.kilograms(value), metadata)
        CsvImportMetric.BODY_FAT -> BodyFatRecord(
            time,
            offset,
            androidx.health.connect.client.units.Percentage(value),
            metadata,
        )
        CsvImportMetric.LEAN_BODY_MASS -> LeanBodyMassRecord(time, offset, Mass.kilograms(value), metadata)
        CsvImportMetric.BONE_MASS -> BoneMassRecord(time, offset, Mass.kilograms(value), metadata)
        CsvImportMetric.BODY_WATER_MASS -> BodyWaterMassRecord(time, offset, Mass.kilograms(value), metadata)
        CsvImportMetric.HEIGHT -> HeightRecord(time, offset, Length.meters(value), metadata)
        CsvImportMetric.BASAL_METABOLIC_RATE -> BasalMetabolicRateRecord(
            time,
            offset,
            Power.kilocaloriesPerDay(value),
            metadata,
        )
        // Health Connect models heart rate as a series: one sample, window of one instant.
        CsvImportMetric.HEART_RATE -> HeartRateRecord(
            startTime = time,
            startZoneOffset = offset,
            endTime = time,
            endZoneOffset = offset,
            samples = listOf(HeartRateRecord.Sample(time = time, beatsPerMinute = value.roundToLong())),
            metadata = metadata,
        )
        // Health Connect stores this one as an integer.
        CsvImportMetric.RESTING_HEART_RATE -> RestingHeartRateRecord(
            time = time,
            zoneOffset = offset,
            beatsPerMinute = value.roundToLong(),
            metadata = metadata,
        )
        CsvImportMetric.HEART_RATE_VARIABILITY -> HeartRateVariabilityRmssdRecord(
            time = time,
            zoneOffset = offset,
            heartRateVariabilityMillis = value,
            metadata = metadata,
        )
        CsvImportMetric.OXYGEN_SATURATION -> OxygenSaturationRecord(
            time,
            offset,
            androidx.health.connect.client.units.Percentage(value),
            metadata,
        )
        CsvImportMetric.RESPIRATORY_RATE -> RespiratoryRateRecord(time, offset, value, metadata)
        CsvImportMetric.BODY_TEMPERATURE -> BodyTemperatureRecord(
            time = time,
            zoneOffset = offset,
            temperature = Temperature.celsius(value),
            metadata = metadata,
        )
        CsvImportMetric.BASAL_BODY_TEMPERATURE -> BasalBodyTemperatureRecord(
            time = time,
            zoneOffset = offset,
            temperature = Temperature.celsius(value),
            metadata = metadata,
        )
        CsvImportMetric.BLOOD_GLUCOSE -> BloodGlucoseRecord(
            time = time,
            zoneOffset = offset,
            level = BloodGlucose.millimolesPerLiter(value),
            specimenSource = BloodGlucoseRecord.SPECIMEN_SOURCE_UNKNOWN,
            mealType = MealType.MEAL_TYPE_UNKNOWN,
            relationToMeal = BloodGlucoseRecord.RELATION_TO_MEAL_UNKNOWN,
            metadata = metadata,
        )
        CsvImportMetric.VO2_MAX -> Vo2MaxRecord(
            time = time,
            zoneOffset = offset,
            vo2MillilitersPerMinuteKilogram = value,
            measurementMethod = Vo2MaxRecord.MEASUREMENT_METHOD_OTHER,
            metadata = metadata,
        )
        // The one interval metric. convertCsvRow guarantees end > start.
        CsvImportMetric.STEPS -> {
            val until = checkNotNull(end) { "STEPS needs an end instant." }
            StepsRecord(
                startTime = time,
                startZoneOffset = offset,
                endTime = until.utc,
                endZoneOffset = until.offset,
                count = value.roundToLong(),
                metadata = metadata,
            )
        }
    }

    return CsvConvertedRecord(
        metric = metric,
        targetType = spec.targetType,
        recordType = spec.recordType,
        clientRecordId = clientRecordId,
        instant = time,
        canonicalValue = value,
        record = record,
    )
}

/**
 * The span of an interval row when the file names no end. One minute: not
 * zero (Health Connect needs end > start), not a day. Flutter parity.
 */
private val DefaultIntervalSpan = Duration.ofMinutes(1)

private fun CsvInstant.plusDefaultIntervalSpan(): CsvInstant =
    CsvInstant(utc = utc.plus(DefaultIntervalSpan), offset = offset)

private fun csvMetadata(clientRecordId: String): Metadata =
    Metadata.manualEntry(
        device = Device(type = Device.TYPE_PHONE),
        clientRecordId = clientRecordId,
    )

/**
 * The identity of a CSV-imported record: type and instant, not value, so a
 * corrected re-import replaces the old record. File name, header, unit and
 * mapping are excluded too. For an interval metric the instant is the start.
 *
 * Byte-compatible with the Flutter build: hash input `"<targetType>|<epochMillis>"`,
 * SHA-256 truncated to 16 bytes, shape `csv_<slug>_<hex>`. Do not change.
 */
fun buildCsvClientRecordId(targetType: String, utc: Instant): String {
    val parts = "$targetType|${utc.toEpochMilli()}"
    val bytes = MessageDigest.getInstance("SHA-256").digest(parts.toByteArray(Charsets.UTF_8))
    val digest = buildString(32) {
        for (index in 0 until 16) {
            val byte = bytes[index].toInt() and 0xFF
            append(HexDigits[byte ushr 4])
            append(HexDigits[byte and 0x0F])
        }
    }
    return "${CSV_CLIENT_RECORD_ID_NAMESPACE}_${toCsvStableIdSegment(targetType)}_$digest"
}

private const val HexDigits = "0123456789abcdef"

private val StableIdSegmentRegex = Regex("[^a-z0-9]+")

/** Slugifies [value]: lowercase, non-alphanumeric runs to `_`, trimmed. Matches Flutter. */
private fun toCsvStableIdSegment(value: String): String =
    value.lowercase(Locale.US)
        .replace(StableIdSegmentRegex, "_")
        .trim('_')
        .ifBlank { "record" }

/**
 * The canonical values [metric] would take across [rows], from the real
 * conversion, so the preview shows what will be written.
 */
fun previewCanonicalValues(
    rows: List<List<String>>,
    mapping: CsvImportMapping,
    metric: CsvImportMetric,
): List<Double> {
    val targetType = CsvMetricCatalog[metric]?.targetType ?: return emptyList()

    val values = mutableListOf<Double>()
    rows.forEachIndexed { index, fields ->
        // +2: 1-based, and past the header row.
        val conversion = convertCsvRow(
            row = CsvRow(rowNumber = index + 2, fields = fields),
            mapping = mapping,
        )
        conversion.records
            .filter { it.targetType == targetType }
            .forEach { values += it.canonicalValue }
    }
    return values
}

/**
 * The earliest and latest wall-clock times [rows] resolve to, or null. Wall
 * clock, so it reads the way the file reads. Resolved through the import
 * path, so a day/month mix-up shows up as a backwards span.
 */
fun previewInstantRange(
    rows: List<List<String>>,
    mapping: CsvImportMapping,
): Pair<LocalDateTime, LocalDateTime>? {
    val column = mapping.timestampColumn ?: return null

    var first: LocalDateTime? = null
    var last: LocalDateTime? = null
    for (fields in rows) {
        val text = fields.getOrNull(column.columnIndex)?.trim()?.takeIf { it.isNotEmpty() } ?: continue
        val instant = resolveCsvInstant(text, mapping.dateTime) ?: continue
        val wallClock = LocalDateTime.ofInstant(instant.utc, instant.offset)
        if (first == null || wallClock.isBefore(first)) first = wallClock
        if (last == null || wallClock.isAfter(last)) last = wallClock
    }

    val start = first ?: return null
    val end = last ?: return null
    return start to end
}

private val NonNumericCharsRegex = Regex("""[^0-9,.\-+eE]""")

/**
 * Parses a numeric cell, tolerating a comma decimal and thousands separators.
 * With both separators present, the last one is the decimal point.
 */
fun parseCsvNumber(text: String): Double? {
    var value = text.trim()
    if (value.isEmpty()) return null

    // Strip stray units, currency and spaces.
    value = value.replace(NonNumericCharsRegex, "")
    if (value.isEmpty()) return null

    val lastComma = value.lastIndexOf(',')
    val lastDot = value.lastIndexOf('.')
    if (lastComma >= 0 && lastDot >= 0) {
        value = if (lastComma > lastDot) {
            // 1.234,5 — dot groups, comma decides.
            value.replace(".", "").replaceFirst(",", ".")
        } else {
            // 1,234.5 — comma groups.
            value.replace(",", "")
        }
    } else if (lastComma >= 0) {
        // Only commas: a single one is the decimal point, several are grouping.
        value = if (value.indexOf(',') == lastComma) {
            value.replaceFirst(",", ".")
        } else {
            value.replace(",", "")
        }
    }

    return value.toDoubleOrNull()
}
